### PR TITLE
fix(core) implement ttl property on the in-memory cache

### LINKED ISSUE #1695: kong.tools.database_cache does not pass on ttl value sent by ldap-auth plugin and custom plugins that use the database_cache module
### Summary

The `access` module in the **ldap-auth** plugin caches authenticated credentials using the `kong.tools.database_cache` with
ttl value defined by `conf.cache_ttl`. Full excerpt from the `access` module:

```
local function authenticate(conf, given_credentials)
  local given_username, given_password = retrieve_credentials(given_credentials)
  if given_username == nil then
    return false
  end

  local credential = cache.get_or_set(cache.ldap_credential_key(given_username), function()
    ngx_log(ngx_debug, "[ldap-auth] authenticating user against LDAP server: "..conf.ldap_host..":"..conf.ldap_port)

    local ok, err = ldap_authenticate(given_username, given_password, conf)
    if err ~= nil then ngx_log(ngx_error, err) end
    if not ok then
      return nil
    end
    return {username = given_username, password = given_password}
  end, conf.cache_ttl)

  return credential and credential.password == given_password, credential
end
```

However the ttl value is not use inside `kong.tools.database_cache` when the element is cached. Instead `exptime = 0` is sent to the cache as `ttl` value that in turn is translated into `-1` (=never expires) inside the cache implementation.

The problem is that the credentials never expire from the cache.

Suggested solution is of course to pass on the `ttl`  value to the cache.
### Steps To Reproduce

The issue was discovered while implementing a custom plugin. While looking at how existing plugins in Kong use the `kong.tools.database_cache` with ttl values we found that this was not only an issue with our custom plugin, but a more general issue.
### Additional Details & Logs
- Kong version (`$ kong version`)
  - 0.9.2
- Kong debug-level startup logs (`$ kong start --vv`)
- Kong error logs (`<KONG_PREFIX>/logs/error.log`)
- Kong configuration (registered APIs/Plugins & configuration file)
- Operating System

### PR DESCRIPTION
**Pull request by Tieske**
### Summary

 - implements `ttl` property on the in-memory cache
 - fixes #1695, closes #1733 (ldap credentials not expiring)
 - separates the 'shm' and 'lua' based caches in the test setup
 - removes closures for the `get_or_set` method as a performance improvement

This is just a first step for #1835

### CODE REVIEW

**Comment by thibaultcha** [comment]
It seems this does not pass linting? Could you take a look at it?

**Comment by Tieske** [comment]
@thibaultcha just updated title and description, now includes the whole caching update so far. Please have a look.

File: kong/tools/database_cache.lua [hunk 1]
```diff
@@ -49,28 +52,62 @@ end
...
+  -- check local memory, and verify ttl
+  local value = DATA[key]
+  if value ~= nil then
+    if type(value) ~= "table" or not value[TTL_EXPIRE_KEY] then
+      -- found non-ttl value, just return it
+      return value
+    elseif value[TTL_EXPIRE_KEY] >= now then
+      -- found ttl-based value, within ttl
+      return value.value
+    else
```
Comment (thibaultcha) [inline kong/tools/database_cache.lua]: The previous branch returns, no need for a nesting `else` here.

File: kong/tools/database_cache.lua [hunk 2]
```diff
@@ -49,28 +52,62 @@ end
...
+  -- nothing found yet, get it from Shared Dictionary
+  value = _M.sh_get(key)
+  if not value then
+    -- nothing found
+    return nil
+  else
+    value = cjson.decode(value)
+    DATA[key] = value  -- store in memory, so we don't need to deserialize next time
+
+    if type(value) ~= "table" or not value[TTL_EXPIRE_KEY] then
```
Comment (thibaultcha) [inline kong/tools/database_cache.lua]: imo, mixing tables and plain values leads to really confusing code, and forces us to deal with such corner-cases. Ex: checking for return type, checking for `expire`, returning `value.value` or just `value`... All of that not only is confusing to the reader and can easily result in a "disaster waiting to happen", but it also prevents proper branch optimization, and just plain old optimization, for that matter (using `cjson.encode()` when the stored values are plain scalars...).

File: kong/tools/database_cache.lua [hunk 3]
```diff
@@ -83,7 +120,16 @@ function _M.delete_all()
...
+-- @param cb callback function. If no data is found under `key`, then the callback
+-- is called with the additional parameters. The result from the callback is
+-- then stored in the cache, and returned.
+-- @param ... the additional parameters passed to `cb`
+-- @return the (newly) cached value
+function _M.get_or_set(key, ttl, cb, ...)
+
   -- Try to get the value from the cache
   local value = _M.get(key)
   if value then return value end
```
Comment (thibaultcha) [inline kong/tools/database_cache.lua]: Shouldn't this be `if value ~= nil` again?

File: kong/tools/database_cache.lua [hunk 4]
```diff
@@ -109,9 +155,9 @@ function _M.get_or_set(key, cb)
   value = _M.get(key)
   if not value then
```
Comment (thibaultcha) [inline kong/tools/database_cache.lua]: ditto: `value ~= nil`

File: kong/api/routes/cache.lua [hunk 5]
```diff
@@ -1,18 +1,60 @@
...
+})
+
+-- build string with valid cache names (for error mesage above)
+for name in pairs(caches) do cache_names[#cache_names+1] = "'"..name.."'" end
+table.sort(cache_names)  -- make order deterministic, for test purposes
+cache_names = table.concat(cache_names, ", ")
+
+
+local function get_cache(self)
+  self.params.cache = caches[ngx.req.get_uri_args()["cache"] or "lua"]
```
Comment (thibaultcha) [inline kong/api/routes/cache.lua]: We are storing a function inside Lapis' `self.params` table? That just doesn't seem right... Also, we are calling `ngx.req.get_uri_args()` when the content of the `cache` URI argument is already in `self.params.cache`, isn't it?

How about storing it in `ngx.ctx` or even `self.kong_cache`?

File: kong/api/routes/cache.lua [hunk 6]
```diff
@@ -1,18 +1,60 @@
 local responses = require "kong.tools.responses"
 local cache = require "kong.tools.database_cache"
 
+local cache_names = {}
```
Comment (thibaultcha) [inline kong/api/routes/cache.lua]: This doesn't need to be in the plugin scope. How about enclosing it with `do ... end`?
```lua
local caches

do
  local cache_names = {}
  for name in pairs(caches) do 
    cache_names[#cache_names + 1] = "'" .. name .. "'" 
  end
  
  -- make order deterministic, for test purposes
  table.sort(cache_names)

  cache_names = table.concat(cache_names, ", ")

  caches = setmetatable({
    ...
  })
end
```

File: kong/core/plugins_iterator.lua [hunk 7]
```diff
@@ -5,6 +5,28 @@ local pl_tablex = require "pl.tablex"
...
+
+  if #rows > 0 then
+    for _, row in ipairs(rows) do
+      if api_id == row.api_id and consumer_id == row.consumer_id then
+        return row
+      end
+    end
+  else
+    -- insert a cached value to not trigger too many DB queries.
+    return {null = true}
```
Comment (thibaultcha) [inline kong/core/plugins_iterator.lua]: The same way we have a `local empty = {}` at the module level here, should we take this as an opportunity to add:

```lua
local t_empty = {}
local t_null = {}
```

?

File: kong/api/routes/cache.lua (hunk 6 above)
Comment (Tieske) [inline kong/api/routes/cache.lua]: done

File: kong/core/plugins_iterator.lua (hunk 7 above)
Comment (Tieske) [inline kong/core/plugins_iterator.lua]: probably a good idea, but postponing this for now

File: kong/tools/database_cache.lua [hunk 8]
```diff
@@ -49,28 +52,62 @@ end
...
+      value = value,
+      [TTL_EXPIRE_KEY] = gettime() + exptime,
+    }
+  end
+
   DATA[key] = value
 
   -- Save into Shared Dictionary
-  local _, err = _M.sh_set(key, cjson.encode(value))
+  local _, err = _M.sh_set(key, cjson.encode(value), exptime)
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: done

File: kong/tools/database_cache.lua (hunk 1 above)
Comment (Tieske) [inline kong/tools/database_cache.lua]: done

File: kong/tools/database_cache.lua [hunk 9]
```diff
@@ -49,28 +52,62 @@ end
...
+      DATA[key] = nil
+      -- value = nil  -- commented out because it triggers a linter error
     end
   end
-
-  return value
+  
+  -- nothing found yet, get it from Shared Dictionary
+  value = _M.sh_get(key)
+  if not value then
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: done

File: kong/tools/database_cache.lua [hunk 10]
```diff
@@ -49,28 +52,62 @@ end
...
+    -- nothing found
+    return nil
+  else
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: done

File: kong/tools/database_cache.lua [hunk 11]
```diff
@@ -49,28 +52,62 @@ end
...
+    value = cjson.decode(value)
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: applied it

File: kong/tools/database_cache.lua [hunk 12]
```diff
@@ -49,28 +52,62 @@ end
...
+    DATA[key] = value  -- store in memory, so we don't need to deserialize next time
+
+    if type(value) ~= "table" or not value[TTL_EXPIRE_KEY] then
+      -- found non-ttl value, just return it
+      return value
+    else
```
Comment (Tieske) [inline kong/tools/database_cache.lua]: done

File: kong/tools/database_cache.lua (hunk 2 above)
Comment (Tieske) [inline kong/tools/database_cache.lua]: I see what you mean, but postponing this change for now

File: kong/tools/database_cache.lua (hunk 3 above)
Comment (Tieske) [inline kong/tools/database_cache.lua]: good catch, fixed

File: kong/api/routes/cache.lua (hunk 5 above)
Comment (Tieske) [inline kong/api/routes/cache.lua]: fixed

**Pull request merged**